app/src/main/java/com/fstyle/structure_android/utils/Validator.java

```diff
@@ -0,0 +1,293 @@
+                    Log.e(TAG, "validate: ", e);
+                } catch (InvocationTargetException e) {
+                    Log.e(TAG, "validate: ", e);
+                }
+                if (mErrorTraces.indexOfKey(type) != -1) {
+                    mMessageInvalids.add(rule.message());
+                }
+            }
+        }
+    }
+
+    public <T extends BaseModel> Validator prepare(T model) {
+        try {
+            mModelCache = model.clone();
+        } catch (CloneNotSupportedException e) {
+            Log.e(TAG, "prepare: ", e);
+        }
+        return this;
+    }
+
+    public <T extends BaseModel> void validate(T model, boolean onlyValidateChange) {
+        if (model == null || mModelCache == null) {
+            return;
+        }
+
+        for (Field field : model.getClass().getDeclaredFields()) {
+            Validation annotation = field.getAnnotation(Validation.class);
+            if (annotation == null) {
+                continue;
+            }
+            Rule[] rules = annotation.value();
+            Optional optional = field.getAnnotation(Optional.class);
+            boolean isOptional = optional != null;
+            field.setAccessible(true);
+
+            try {
+                Object cache = field.get(mModelCache);
+                Object real = field.get(model);
+                // detect when data has changed only if onlyValidateTheChange is true.
+                // Otherwise, always run validation on all of fields .
+                if (!onlyValidateChange || (cache == null && real != null) || (cache != null
+                        && cache.equals(real))) {
+                    validate(real, rules, isOptional);
+                }
+            } catch (IllegalAccessException e) {
+                Log.e(TAG, "validate: ", e);
+            }
+            try {
+                mModelCache = model.clone();
+            } catch (CloneNotSupportedException e) {
+                mModelCache = null;
+                Log.e(TAG, "validate: ", e);
+            }
+        }
+    }
+
+    @StringRes
+    public int getErrorByTypeInt(@ValidType int type) {
+        int index = mErrorTraces.indexOfValue(type);
+        if (index == -1) {
+            return R.string.empty;
+        }
+        return mErrorTraces.keyAt(index);
+    }
+
+    public String getErrorByType(@ValidType int type) {
+        int index = mErrorTraces.indexOfValue(type);
+        if (index == -1 || mErrorTraces.keyAt(index) == -1) {
+            return null;
+        }
+        return mContext.getString(mErrorTraces.keyAt(index));
+    }
+
+    public void initNGWordPattern() {
+        Observable.create(new Observable.OnSubscribe<Pattern>() {
+            @Override
+            public void call(Subscriber<? super Pattern> subscriber) {
+                try {
+                    BufferedReader reader = new BufferedReader(
+                            new InputStreamReader(mContext.getAssets().open("ng-word")));
```

Comment:
ERROR: [I18N][DM_DEFAULT_ENCODING] Found reliance on default encoding in com.fstyle.structure_android.utils.Validator$3.call(Subscriber): new java.io.InputStreamReader(InputStream)
http://findbugs.sourceforge.net/bugDescriptions.html#DM_DEFAULT_ENCODING